File: Legacy_Code_Conv/SODA/Modernization/backend/src/core/database/database.config.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '../config/config.service';
import { LoggerService } from '../logger/logger.service';
import { TypeOrmModuleOptions } from '@nestjs/typeorm';

@Injectable()
export class DatabaseConfig {
  constructor(
    private readonly config: ConfigService,
    private readonly logger: LoggerService
  ) {}

  /**
   * Gets TypeORM configuration for the main database connection
   */
  getTypeOrmConfig(): TypeOrmModuleOptions {
    const config: TypeOrmModuleOptions = {
      type: 'postgres',
      host: this.config.get('DB_HOST', 'localhost'),
      port: this.config.getNumber('DB_PORT', 5432),
      username: this.config.get('DB_USERNAME', 'postgres'),
      password: this.config.get('DB_PASSWORD', ''),
      database: this.config.get('DB_DATABASE', 'ariesone'),
      entities: ['dist/**/*.entity{.ts,.js}'],
      migrations: ['dist/migrations/*{.ts,.js}'],
      synchronize: this.config.get('NODE_ENV', 'development') === 'development',
      logging: this.config.getBoolean('DB_LOGGING', false),
      logger: 'advanced-console',
      maxQueryExecutionTime: 1000, // Log slow queries (>1s)
      cache: {
        type: 'redis',
        options: {
          host: this.config.get('REDIS_HOST', 'localhost'),
          port: this.config.getNumber('REDIS_PORT', 6379),
          password: this.config.get('REDIS_PASSWORD', ''),
          db: this.config.getNumber('REDIS_DB', 0)
        },
        duration: 60000 // 1 minute
      },
      extra: {
        max: this.config.getNumber('DB_POOL_MAX', 20), // Max connections
        connectionTimeoutMillis: 3000, // Connection timeout
        idleTimeoutMillis: 30000, // How long a connection can be idle
        ssl: this.config.getBoolean('DB_SSL', false)
          ? {
              rejectUnauthorized: false
            }
          : undefined
      }
    };

    this.logger.debug('Database configuration loaded', {
      host: config.host,
      port: config.port,
      database: config.database,
      maxConnections: config.extra?.max
    });

    return config;
  }
}

File: Legacy_Code_Conv/SODA/Modernization/backend/app/core/database/database_config.py
"""Database configuration module for AriesOne SaaS platform."""

from typing import Any, Dict, Optional
from pydantic_settings import BaseSettings
from functools import lru_cache


class DatabaseSettings(BaseSettings):
    """Database configuration settings."""
    
    # PostgreSQL settings
    DB_HOST: str = "localhost"
    DB_PORT: int = 5432
    DB_USERNAME: str = "postgres"
    DB_PASSWORD: str = ""
    DB_DATABASE: str = "ariesone"
    DB_SSL: bool = False
    DB_POOL_SIZE: int = 20
    DB_POOL_OVERFLOW: int = 10
    DB_TIMEOUT: int = 30
    DB_POOL_RECYCLE: int = 300
    DB_ECHO: bool = False
    
    # Redis settings
    REDIS_HOST: str = "localhost"
    REDIS_PORT: int = 6379
    REDIS_PASSWORD: str = ""
    REDIS_DB: int = 0
    REDIS_SSL: bool = False
    
    # Cache settings
    CACHE_TTL: int = 300  # 5 minutes
    
    class Config:
        """Pydantic configuration class."""
        env_file = ".env"
        case_sensitive = True

    @property
    def database_url(self) -> str:
        """Get the database URL for SQLAlchemy."""
        return (
            f"postgresql+asyncpg://{self.DB_USERNAME}:{self.DB_PASSWORD}"
            f"@{self.DB_HOST}:{self.DB_PORT}/{self.DB_DATABASE}"
        )
    
    @property
    def redis_url(self) -> str:
        """Get the Redis URL."""
        protocol = "rediss" if self.REDIS_SSL else "redis"
        auth = f":{self.REDIS_PASSWORD}@" if self.REDIS_PASSWORD else ""
        return f"{protocol}://{auth}{self.REDIS_HOST}:{self.REDIS_PORT}/{self.REDIS_DB}"

    def get_pool_settings(self) -> Dict[str, Any]:
        """Get database pool settings."""
        return {
            "pool_size": self.DB_POOL_SIZE,
            "max_overflow": self.DB_POOL_OVERFLOW,
            "pool_timeout": self.DB_TIMEOUT,
            "pool_recycle": self.DB_POOL_RECYCLE,
            "pool_pre_ping": True,
            "echo": self.DB_ECHO,
            "ssl": self.DB_SSL
        }


@lru_cache()
def get_database_settings() -> DatabaseSettings:
    """Get cached database settings instance."""
    return DatabaseSettings()
